presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -189,6 +219,143 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+
+            return decorrelatedOrderingScheme
+                    .map(orderingScheme -> {
+                        // ordering symbols are present - rewrite TopN to TopNRowNumberNode partitioned by constant symbols
+                        TopNRowNumberNode topNRowNumberNode = new TopNRowNumberNode(
+                                node.getId(),
+                                decorrelatedChildNode,
+                                new Specification(
+                                        ImmutableList.copyOf(childDecorrelationResult.symbolsToPropagate),
+                                        Optional.of(orderingScheme)),
+                                symbolAllocator.newSymbol("row_number", BIGINT),
+                                toIntExact(node.getCount()),
+                                false,
+                                Optional.empty());
+
+                        return Optional.of(new DecorrelationResult(
+                                topNRowNumberNode,
+                                childDecorrelationResult.symbolsToPropagate,
+                                childDecorrelationResult.correlatedPredicates,
+                                childDecorrelationResult.correlatedSymbolsMapping,
+                                node.getCount() == 1));
+                    })
+                    .orElseGet(() -> {
+                        // no ordering symbols are left - rewrite TopN to RowNumberNode partitioned by constant symbols
+                        RowNumberNode rowNumberNode = new RowNumberNode(
+                                node.getId(),
+                                decorrelatedChildNode,
+                                ImmutableList.copyOf(childDecorrelationResult.symbolsToPropagate),
+                                symbolAllocator.newSymbol("row_number", BIGINT),
+                                Optional.of(toIntExact(node.getCount())),
+                                Optional.empty());
+
+                        return Optional.of(new DecorrelationResult(
+                                rowNumberNode,
+                                childDecorrelationResult.symbolsToPropagate,
+                                childDecorrelationResult.correlatedPredicates,
+                                childDecorrelationResult.correlatedSymbolsMapping,
+                                node.getCount() == 1));
+                    });
+        }
+
+        private Optional<OrderingScheme> decorrelateOrderingScheme(OrderingScheme orderingScheme, Set<Symbol> constantSymbols)
+        {
+            // remove local and remote constant sort symbols from the OrderingScheme
+            ImmutableList.Builder<Symbol> nonConstantOrderBy = ImmutableList.builder();
+            ImmutableMap.Builder<Symbol, SortOrder> nonConstantOrderings = ImmutableMap.builder();
+            for (Symbol symbol : orderingScheme.getOrderBy()) {
+                if (!constantSymbols.contains(symbol) && !correlation.contains(symbol)) {
```

Comment:
Could you point me to actual query that test `!correlation.contains(symbol)`?

Reply:
That would be any query with `ORDER BY` a local symbol - the symbol passes the condition.
An example with a symbol that doesn't pass:
```
assertions.assertFails(
                "SELECT * " +
                        "FROM (VALUES 1, 2, 3, null) outer_relation(id) " +
                        "LEFT JOIN LATERAL " +
                        "(SELECT value FROM (VALUES 'c', 'a', 'b') inner_relation(value) " +
                        "   WHERE outer_relation.id = 3 ORDER BY outer_relation.id LIMIT 2) " +
                        "ON TRUE",
                UNSUPPORTED_CORRELATED_SUBQUERY_ERROR_MSG);
```
(I need to fix the comment for this one).

Reply:
I mean if we could have a test that shows improvement vs using mapper. I though the one with "corr_sym = 3 order by corr_sym" would be such test

Reply:
Yes, this one matches too.
The problem is that for any query that has ordering by a correlation symbol, there 'appears' an identity projection on this symbol below `TopNNode`.
And so far, `ProjectNode`'s assignments aren't decorrelated. So currently this path cannot be tested, because the query will fail with `Given correlated subquery not supported`, although `TopNNode` itself decorrelates.
That's why I intended to decorrelate Projections, as they are the real blocker.